exonum/src/runtime/migrations.rs

```diff
@@ -0,0 +1,485 @@
+    FutureStartVersion {
+        /// Maximum supported version.
+        max_supported_version: Version,
+    },
+    /// The start version falls in the supported lower / upper bounds on versions,
+    /// but is not supported itself. This can be the case, e.g., for pre-releases.
+    UnsupportedStart,
+    // TODO: probably need custom errors.
+}
+
+/// Linearly ordered migrations.
+///
+/// FIXME: more details, examples
+///
+/// # Limitations
+///
+/// Special care must be taken to support pre-release versions (i.e., versions like `0.2.0-pre.2` or
+/// `1.2.34-rc.5`). As per the semver specification:
+///
+/// - Any pre-release version is lesser than a corresponding version without a pre-release suffix
+/// - Pre-release versions with the same base triple are ordered alphabetically by their suffixes
+///
+/// The support of prerelease versions needs to be explicitly enabled by using the
+/// [`with_prereleases`](#method.with_prereleases) constructor. Otherwise, a prerelease mentioned
+/// in the builder stage will lead to a panic, and [`select`](#method.select) will return an error
+/// if a prerelease is specified as a `start_version`.
+#[derive(Debug)]
+pub struct LinearMigrations {
+    min_start_version: Option<Version>,
+    latest_version: Version,
+    scripts: BTreeMap<Version, MigrationScript>,
+    support_prereleases: bool,
+}
+
+impl LinearMigrations {
+    /// Creates a new set of migrations with the specified latest supported version.
+    pub fn new(latest_version: Version) -> Self {
+        assert!(
+            !latest_version.is_prerelease(),
+            "Prerelease versions require using `with_prereleases` constructor"
+        );
+        Self {
+            support_prereleases: false,
+            ..Self::with_prereleases(latest_version)
+        }
+    }
+
+    /// Creates a new set of migrations with the specified latest supported version.
+    ///
+    /// Unlike the `new` constructor, this one indicates that prerelease versions are allowed
+    /// in the list of migrations and as the start version.
+    pub fn with_prereleases(latest_version: Version) -> Self {
+        Self {
+            min_start_version: None,
+            latest_version,
+            scripts: BTreeMap::new(),
+            support_prereleases: true,
+        }
+    }
+
+    fn check_prerelease(&self, version: &Version) {
+        if !self.support_prereleases {
+            assert!(
+                !version.is_prerelease(),
+                "Prerelease versions require using `with_prereleases` constructor"
+            );
+        }
+    }
+
+    /// Signals to return an error if the starting version is less than the specified version.
+    pub fn forget_before(mut self, version: Version) -> Self {
+        self.check_prerelease(&version);
+        self.min_start_version = Some(version);
+        self
+    }
+
+    /// Adds a migration script at the specified version.
+    pub fn migrate<F>(mut self, version: Version, script: F) -> Self
```

Comment:
Hm, IMHO this name is confusing, I'd name it `add_migration_script` or `add_migration`.